lib/internal/test_runner/tap_lexer.js

```diff
@@ -0,0 +1,589 @@
+'use strict';
+
+const TokenKind = {
+  EOF: 'EOF',
+  EOL: 'EOL',
+  NUMERIC: 'Numeric',
+  LITERAL: 'Literal',
+  KEYWORD: 'Keyword',
+  WHITESPACE: 'Whitespace',
+  COMMENT: 'Comment',
+  DASH: 'Dash',
+  PLUS: 'Plus',
+  HASH: 'Hash',
+  ESCAPE: 'Escape',
+
+  // TAP tokens
+  TAP: 'TAPKeyword',
+  TAP_VERSION: 'VersionKeyword',
+  TAP_PLAN: 'PlanKeyword',
+  TAP_TEST_OK: 'TestOkKeyword',
+  TAP_TEST_NOTOK: 'TestNotOkKeyword',
+  TAP_YAML: 'YamlKeyword',
+  TAP_PRAGMA: 'PragmaKeyword',
+};
+
+class Token {
+  constructor({ kind, value, stream }) {
+    this.kind = kind;
+    this.value = value;
+    this.location = {
+      line: stream.line,
+      column: Math.max(stream.column - ('' + value).length + 1, 0),
+      start: stream.pos - ('' + value).length - 1, // zero based
+      end: stream.pos - 2, // zero based
+    };
+  }
+}
+
+class InputStream {
+  constructor(input) {
+    this.input = input;
+    this.pos = 0;
+    this.column = 0;
+    this.line = 1;
+  }
+
+  eof() {
+    return this.peek() === undefined;
+  }
+
+  peek(offset = 0) {
+    return this.input[this.pos + offset];
```

Comment:
can this support a stream?
otherwise the test runner will wait for each file to finish before it starts reporting

Reply:
Interesting! Adding this to my todo